nucliadb_node/nucliadb_node/pull.py

```diff
@@ -61,369 +51,65 @@
-                    )
-            except Exception:
-                logger.exception(
-                    "Could not garbage collect", extra={"shard": self._shard_id}
-                )
 
 
 class Worker:
-    subscriptions: List[Subscription]
+    subscriptions: list[Subscription]
     storage: Storage
 
-    def __init__(self, writer: Writer, node: str):
+    def __init__(
+        self,
+        writer: Writer,
+        node: str,
+        nats_connection_manager: NatsConnectionManager,
+    ):
         self.writer = writer
         self.subscriptions = []
         self.node = node
-        self.publisher = IndexedPublisher()
-        self.load_seqid()
-        self.brain: Optional[Resource] = None
-
-        self.shard_managers: dict[str, ShardManager] = {}
-        # right now, only allow one gc at a time but
-        # can be expanded to allow multiple if we want with a semaphore
-        self.gc_lock = asyncio.Semaphore(1)
+        self.indexer = ConcurrentShardIndexer(self.writer)
+        self.nats_connection_manager = nats_connection_manager
 
     async def initialize(self):
         self.storage = await get_storage(service_name=SERVICE_NAME)
-        await self.publisher.initialize()
-        await self.subscriber_initialize()
-        await self.garbage_collect_all()
+        await self.indexer.initialize()
+        await self.setup_nats_subscription()
 
     async def finalize(self):
-        await self.publisher.finalize()
-        await self.subscriber_finalize()
-
-        await self.storage.finalize()
-
-    async def disconnected_cb(self):
-        logger.info("Got disconnected from NATS!")
-
-    async def reconnected_cb(self):
-        # See who we are connected to on reconnect
-        logger.warning(
-            f"Got reconnected to NATS {self.nc.connected_url}. Attempting reconnect"
-        )
-        await self.drain_subscriptions()
-        await self.subscribe()
-
-    async def error_cb(self, e):
-        errors.capture_exception(e)
-        logger.error(
-            "There was an error on the worker, check sentry: {}".format(e),
-            exc_info=True,
-        )
-
-    async def closed_cb(self):
-        logger.info("Connection is closed on NATS")
-
-    async def subscriber_initialize(self):
-        options = {
-            "error_cb": self.error_cb,
-            "closed_cb": self.closed_cb,
-            "reconnected_cb": self.reconnected_cb,
-        }
-
-        if indexing_settings.index_jetstream_auth:
-            options["user_credentials"] = indexing_settings.index_jetstream_auth
-
-        if len(indexing_settings.index_jetstream_servers) > 0:
-            options["servers"] = indexing_settings.index_jetstream_servers
-
-        self.nc = await nats.connect(**options)
-        self.js = get_traced_jetstream(self.nc, SERVICE_NAME)
-        logger.info(f"Nats: Connected to {indexing_settings.index_jetstream_servers}")
-        await self.subscribe()
+        await self.indexer.finalize()
```

Comment:
Typically you want to finalize in the reverse order of initialization.
Otherwise you may be pushing new messages to an indexed that has finished / is finishing and that may cause errors.

Reply:
it was a comment from my review 2 weeks ago, yup